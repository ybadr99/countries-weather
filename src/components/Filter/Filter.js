import React, { useState, useEffect } from 'react';
import { FaArrowUp, FaArrowDown } from 'react-icons/fa';
import { useDispatch } from 'react-redux';
import data from '../../data.json';
import { setCountries } from '../../redux/countries/countriesSlice';
import './Filter.scss';

const Filter = () => {
  const [isOpen, setIsOpen] = useState(false);
  const [selectedRegion, setSelectedRegion] = useState('Filter by region');
  const [searchInput, setSearchInput] = useState('');

  const allRegions = ['All Regions', ...new Set(data.map((el) => el?.region))];
  const handleRegion = (e, region) => {
    e.stopPropagation();
    setSelectedRegion(region);
    setIsOpen(false);
  };

  const handleCountryChange = (e) => {
    setSearchInput(e.target.value);
  };

  const dispatch = useDispatch();

  useEffect(() => {
    if (selectedRegion !== 'Filter by region') {
      if (selectedRegion === 'All Regions') {
        dispatch(setCountries(data));
      } else {
        const selectedRegionArray = data.filter(
          (el) => el?.region === selectedRegion,
        );
        dispatch(setCountries(selectedRegionArray));
      }
    }
    if (searchInput === '') {
      dispatch(setCountries(data));
    } else if (searchInput) {
      // eslint-disable-next-line max-len
      const filteredCountry = data.filter((el) => el.name.toLowerCase().includes(searchInput.toLowerCase()));
      dispatch(setCountries(filteredCountry));
    }
  }, [
    selectedRegion,
    setSelectedRegion,
    searchInput,
    setSearchInput,
    dispatch,
  ]);

  return (
    <div className="search">
      <input
        type="text"
        placeholder="Search country"
        onChange={handleCountryChange}
        value={searchInput}
      />
      <button
        type="button"
        className="filter"
        onClick={() => setIsOpen(!isOpen)}
      >
        <div>
          {selectedRegion}
          {isOpen ? <FaArrowUp /> : <FaArrowDown />}
        </div>
        {isOpen && (
          <div className="regions">
            {allRegions.map((region, idx) => (
              <button
                type="button"
                onClick={(e) => handleRegion(e, region)}
                key={idx}
              >
                {region}
              </button>
            ))}
          </div>
        )}
      </button>
    </div>
  );
};

export default Filter;
