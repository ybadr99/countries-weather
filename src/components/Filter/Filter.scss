@import '../../style/variables.scss';

.filter {
  padding: 1rem;
  color: $white;
  position: relative;
  background-color: $dark-blue;
  width: 25%;
  border-radius: 0.2rem;
  border: none;
  cursor: pointer;
  margin: 20px 0;

  div {
    display: flex;
    align-items: center;
    gap: 1rem;
    justify-content: space-between;
    color: $white;

    &.regions {
      width: 100%;
      position: absolute;
      height: 10rem;
      overflow-y: scroll;
      top: 3.5rem;
      display: flex;
      gap: 0;
      flex-direction: column;
      left: 0;
      z-index: 99;

      button {
        width: 100%;
        padding: 1rem;
        color: $white;
        border: none;
        flex: 1;
        cursor: pointer;
        z-index: 99;
        text-align: left;
        background-color: $dark-blue;

        &:hover {
          background-color: $dark-gray;
        }
      }
    }
  }
}
