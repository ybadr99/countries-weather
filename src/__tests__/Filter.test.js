import React from 'react';
import { Provider } from 'react-redux';
import { configureStore, combineReducers } from '@reduxjs/toolkit';
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer';
import Filter from '../components/Filter/Filter';
import countriesReducer from '../redux/countries/countriesSlice';
import '@testing-library/jest-dom/extend-expect'; // Corrected import

describe('Filter component', () => {
  const countries = [
    {
      name: 'Afghanistan',
      topLevelDomain: ['.af'],
    },
    {
      name: 'Ã…land Islands',
      topLevelDomain: ['.ax'],
    },
  ];

  const rootReducer = combineReducers({
    countries: countriesReducer,
  });

  const store = configureStore({
    reducer: rootReducer,
    preloadedState: {
      countries: {
        countries,
        status: 'success',
        error: null,
      },
    },
  });

  it('renders the countries correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Filter />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
