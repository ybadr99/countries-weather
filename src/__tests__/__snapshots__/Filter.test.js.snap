// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component renders the countries correctly 1`] = `
<button
  className="filter"
  onClick={[Function]}
  type="button"
>
  <div>
    Filter by region
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
      />
    </svg>
  </div>
</button>
`;
